File: app/src/main/java/com/example/criminalintent/DatePickerFragment.java
package com.example.criminalintent;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.DatePicker;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class DatePickerFragment extends DialogFragment {

    private static final String ARG_DATE = "date";
    private static final String ARG_REQUEST_KEY = "request_key";
    private static final String ARG_RESULT_KEY = "result_key";
    private DatePicker mDatePicker;

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Date date = SerializationCompat.getSerializable(getArguments(), ARG_DATE, Date.class);
        if (date == null) {
            date = new Date();
        }

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DAY_OF_MONTH);

        View view = LayoutInflater.from(getActivity()).inflate(R.layout.dialog_date, null);
        mDatePicker = (DatePicker) view.findViewById(R.id.dialog_date_picker);
        mDatePicker.init(year, month, day, null);

        return new AlertDialog.Builder(getActivity())
                .setView(view)
                .setTitle(R.string.date_picker_title)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    int pickedYear = mDatePicker.getYear();
                    int pickedMonth = mDatePicker.getMonth();
                    int pickedDay = mDatePicker.getDayOfMonth();
                    Date pickedDate = new GregorianCalendar(pickedYear, pickedMonth, pickedDay).getTime();
                    sendResult(pickedDate);
                })
                .create();
    }

    public static DatePickerFragment newInstance(Date date, String requestKey, String resultKey) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_DATE, date);
        args.putString(ARG_REQUEST_KEY, requestKey);
        args.putString(ARG_RESULT_KEY, resultKey);
         DatePickerFragment fragment = new DatePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private void sendResult(Date date) {
        Bundle arguments = getArguments();
        if (arguments == null) {
            return;
        }

        String requestKey = arguments.getString(ARG_REQUEST_KEY);
        String resultKey = arguments.getString(ARG_RESULT_KEY);
        if (requestKey == null || resultKey == null) {
            return;
        }

        Bundle result = new Bundle();
        result.putSerializable(resultKey, date);
        getParentFragmentManager().setFragmentResult(requestKey, result);
    }
}

File: app/src/main/java/com/bignerdranch/android/criminalintent/DatePickerFragment.kt
package com.bignerdranch.android.criminalintent

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import java.util.Calendar
import java.util.Date
import java.util.GregorianCalendar

private const val ARG_DATE = "date"

class DatePickerFragment : DialogFragment() {

    companion object {
        const val EXTRA_DATE = "com.bignerdranch.android.criminalintent.date"

        fun newInstance(date: Date): DatePickerFragment {
            val args = Bundle().apply {
                putSerializable(ARG_DATE, date)
            }

            return DatePickerFragment().apply {
                arguments = args
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val v = inflater.inflate(R.layout.dialog_date, container, false)

        val date = arguments?.getSerializable(ARG_DATE) as Date
        val calendar = Calendar.getInstance()
        calendar.time = date
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        val datePicker = v.findViewById<DatePicker>(R.id.dialog_date_picker)
        datePicker.init(year, month, day, null)

        val okButton = v.findViewById<Button>(R.id.dialog_date_ok_button)
        okButton.setOnClickListener {
            val year = datePicker.year
            val month = datePicker.month
            val day = datePicker.dayOfMonth
            val date = GregorianCalendar(year, month, day).time
            sendResult(Activity.RESULT_OK, date)
        }

        return v
    }

    private fun sendResult(resultCode: Int, date: Date) {
        val intent = Intent().apply {
            putExtra(EXTRA_DATE, date)
        }

        if (targetFragment != null) {
            targetFragment?.onActivityResult(targetRequestCode, resultCode, intent)
            dismiss()
        } else {
            activity?.setResult(resultCode, intent)
            activity?.finish()
        }
    }
}
